File: modules/memo-editor/ios/MemoEditorModule.swift
import ExpoModulesCore

/// index번째 문단의 내용이 text이고 종류가 from이면 to로 바꾼다.
struct ParagraphBlockChange: Record {
  @Field var index: Int = 0
  @Field var text: String = ""
  @Field var from: String = ""
  @Field var to: String = ""
}

/// index번째 문단의 내용이 text이고 start부터 length만큼(UTF-16)이 word이면 그 낱말 뒤에 id 두들을 붙인다.
struct DoodleChange: Record {
  @Field var index: Int = 0
  @Field var text: String = ""
  @Field var start: Int = 0
  @Field var length: Int = 0
  @Field var word: String = ""
  @Field var id: String = ""
}

/// 두들 그림 한 겹 (24x24 격자, M·L·C·Q·Z 경로, '#RRGGBB' 색)
struct DoodleOpRecord: Record {
  @Field var d: String = ""
  @Field var fill: String?
  @Field var stroke: String?
  @Field var width: Double?
  @Field var opacity: Double?
  @Field var dx: Double?
  @Field var dy: Double?
}

/// 두들 그림 하나와, 두들을 붙인 낱말을 감싸는 칩의 색
struct DoodleArtRecord: Record {
  @Field var id: String = ""
  @Field var ops: [DoodleOpRecord] = []
  @Field var chipFill: String?
}

public class MemoEditorModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MemoEditor")

    View(MemoEditorView.self) {
      Events(
        "onChangeContent", "onChangeFormat", "onChangeHistory", "onFocusChange", "onLeaveParagraph",
        "onBackspaceWhenEmpty")

      Prop("initialContent") { (view: MemoEditorView, value: String?) in
        view.setInitialContent(value)
      }
      Prop("placeholder") { (view: MemoEditorView, value: String?) in
        view.placeholder = value ?? ""
      }
      Prop("autoFocus") { (view: MemoEditorView, value: Bool?) in
        view.autoFocus = value ?? false
      }
      Prop("fontSize") { (view: MemoEditorView, value: Double?) in
        view.theme.fontSize = CGFloat(value ?? 18)
      }
      Prop("textColor") { (view: MemoEditorView, value: UIColor?) in
        if let value { view.theme.textColor = value }
      }
      Prop("mutedColor") { (view: MemoEditorView, value: UIColor?) in
        if let value { view.theme.mutedColor = value }
      }
      Prop("accentColor") { (view: MemoEditorView, value: UIColor?) in
        if let value { view.theme.accentColor = value }
      }
      Prop("placeholderColor") { (view: MemoEditorView, value: UIColor?) in
        if let value { view.theme.placeholderColor = value }
      }
      Prop("insetHorizontal") { (view: MemoEditorView, value: Double?) in
        view.insetHorizontal = CGFloat(value ?? 20)
      }
      Prop("insetTop") { (view: MemoEditorView, value: Double?) in
        view.insetTop = CGFloat(value ?? 14)
      }
      Prop("accessoryID") { (view: MemoEditorView, value: String?) in
        view.textView.accessoryID = value
      }
      Prop("doodleArt") { (view: MemoEditorView, value: [DoodleArtRecord]?) in
        view.doodleArt = MemoDoodles.parseArt(value ?? [])
      }

      OnViewDidUpdateProps { (view: MemoEditorView) in
        view.didUpdateProps()
      }

      AsyncFunction("focus") { (view: MemoEditorView) in
        view.focus()
      }.runOnQueue(.main)

      AsyncFunction("blur") { (view: MemoEditorView) in
        view.blur()
      }.runOnQueue(.main)

      AsyncFunction("toggleBold") { (view: MemoEditorView) in
        view.toggleInline(\.bold)
      }.runOnQueue(.main)

      AsyncFunction("toggleUnderline") { (view: MemoEditorView) in
        view.toggleInline(\.underline)
      }.runOnQueue(.main)

      AsyncFunction("toggleStrikethrough") { (view: MemoEditorView) in
        view.toggleInline(\.strikethrough)
      }.runOnQueue(.main)

      AsyncFunction("toggleBlock") { (view: MemoEditorView, kind: String) in
        view.toggleBlock(named: kind)
      }.runOnQueue(.main)

      AsyncFunction("setParagraphBlocks") { (view: MemoEditorView, changes: [ParagraphBlockChange]) -> [Bool] in
        view.setParagraphBlocks(changes)
      }.runOnQueue(.main)

      AsyncFunction("setDoodles") { (view: MemoEditorView, changes: [DoodleChange], explicit: Bool) -> [Bool] in
        view.setDoodles(changes, explicit: explicit)
      }.runOnQueue(.main)

      AsyncFunction("removeDoodles") { (view: MemoEditorView) -> Int in
        view.removeDoodles()
      }.runOnQueue(.main)

      AsyncFunction("undo") { (view: MemoEditorView) in
        view.undo()
      }.runOnQueue(.main)

      AsyncFunction("redo") { (view: MemoEditorView) in
        view.redo()
      }.runOnQueue(.main)
    }
  }
}

File: modules/memo-editor/android/src/main/java/expo/modules/memoeditor/MemoEditorModule.kt
package expo.modules.memoeditor

import android.graphics.Color
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

// JS에서는 '#RRGGBB' 문자열로 넘어온다.
private fun parseColor(value: String?): Int? =
  try {
    value?.let { Color.parseColor(it) }
  } catch (error: IllegalArgumentException) {
    null
  }

/** index번째 문단의 내용이 text이고 종류가 from이면 to로 바꾼다. */
class ParagraphBlockChange : Record {
  @Field
  val index: Int = 0

  @Field
  val text: String = ""

  @Field
  val from: String = ""

  @Field
  val to: String = ""
}

/** index번째 문단의 내용이 text이고 start부터 length만큼(UTF-16)이 word이면 그 낱말 뒤에 id 두들을 붙인다. */
class DoodleChange : Record {
  @Field
  val index: Int = 0

  @Field
  val text: String = ""

  @Field
  val start: Int = 0

  @Field
  val length: Int = 0

  @Field
  val word: String = ""

  @Field
  val id: String = ""
}

/** 두들 그림 한 겹 (24x24 격자, M·L·C·Q·Z 경로, '#RRGGBB' 색) */
class DoodleOpRecord : Record {
  @Field
  val d: String = ""

  @Field
  val fill: String? = null

  @Field
  val stroke: String? = null

  @Field
  val width: Double? = null

  @Field
  val opacity: Double? = null

  @Field
  val dx: Double? = null

  @Field
  val dy: Double? = null
}

/** 두들 그림 하나와, 두들을 붙인 낱말을 감싸는 칩의 색 */
class DoodleArtRecord : Record {
  @Field
  val id: String = ""

  @Field
  val ops: List<DoodleOpRecord> = emptyList()

  @Field
  val chipFill: String? = null
}

class MemoEditorModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("MemoEditor")

    View(MemoEditorView::class) {
      Events("onChangeContent", "onChangeFormat", "onChangeHistory", "onFocusChange", "onLeaveParagraph")

      Prop("initialContent") { view: MemoEditorView, value: String? ->
        view.setInitialContent(value)
      }
      Prop("placeholder") { view: MemoEditorView, value: String? ->
        view.placeholder = value ?: ""
      }
      Prop("autoFocus") { view: MemoEditorView, value: Boolean? ->
        view.autoFocus = value ?: false
      }
      Prop("fontSize") { view: MemoEditorView, value: Double? ->
        view.theme.fontSize = (value ?: 18.0).toFloat()
        view.markThemeDirty()
      }
      Prop("textColor") { view: MemoEditorView, value: String? ->
        parseColor(value)?.let { view.theme.textColor = it }
        view.markThemeDirty()
      }
      Prop("mutedColor") { view: MemoEditorView, value: String? ->
        parseColor(value)?.let { view.theme.mutedColor = it }
        view.markThemeDirty()
      }
      Prop("accentColor") { view: MemoEditorView, value: String? ->
        parseColor(value)?.let { view.theme.accentColor = it }
        view.markThemeDirty()
      }
      Prop("placeholderColor") { view: MemoEditorView, value: String? ->
        parseColor(value)?.let { view.theme.placeholderColor = it }
        view.markThemeDirty()
      }
      Prop("insetHorizontal") { view: MemoEditorView, value: Double? ->
        view.insetHorizontal = (value ?: 20.0).toFloat()
      }
      Prop("insetTop") { view: MemoEditorView, value: Double? ->
        view.insetTop = (value ?: 14.0).toFloat()
      }
      // iOS에서 키보드 위에 붙일 컨트롤 바. Android에는 그런 자리가 없어 JS가 키보드를 따라 띄운다.
      Prop("accessoryID") { _: MemoEditorView, _: String? -> }
      Prop("doodleArt") { view: MemoEditorView, value: List<DoodleArtRecord>? ->
        view.setDoodleArt(value ?: emptyList())
      }

      OnViewDidUpdateProps { view: MemoEditorView ->
        view.didUpdateProps()
      }

      AsyncFunction("focus") { view: MemoEditorView ->
        view.focusEditor()
      }
      AsyncFunction("blur") { view: MemoEditorView ->
        view.blurEditor()
      }
      AsyncFunction("toggleBold") { view: MemoEditorView ->
        view.toggleInline(InlineStyle.BOLD)
      }
      AsyncFunction("toggleUnderline") { view: MemoEditorView ->
        view.toggleInline(InlineStyle.UNDERLINE)
      }
      AsyncFunction("toggleStrikethrough") { view: MemoEditorView ->
        view.toggleInline(InlineStyle.STRIKETHROUGH)
      }
      AsyncFunction("toggleBlock") { view: MemoEditorView, kind: String ->
        view.toggleBlock(kind)
      }
      AsyncFunction("setParagraphBlocks") { view: MemoEditorView, changes: List<ParagraphBlockChange> ->
        view.setParagraphBlocks(changes)
      }
      AsyncFunction("setDoodles") { view: MemoEditorView, changes: List<DoodleChange>, explicit: Boolean ->
        view.setDoodles(changes, explicit)
      }
      AsyncFunction("removeDoodles") { view: MemoEditorView ->
        view.removeDoodles()
      }
      AsyncFunction("undo") { view: MemoEditorView ->
        view.undo()
      }
      AsyncFunction("redo") { view: MemoEditorView ->
        view.redo()
      }
    }
  }
}
